Add ability to create custom pipeline

pacmessica opened the pull request:
https://www.notion.so/figmentnetworks/b47c0f8b935741d8ac28717ddeea2038?v=5e4e1d3854094b788ae2b612329f27e8&p=2523fa1abb83441d889fde1851b455ec

sosedoff commented:
Are there any info/tickets to read up on these changes? I've not been following the engine project.

pacmessica commented:
@sosedoff For sure! I updated the PR description with the notion link. 

I was just updating the README to explain the usage before moving the notion ticket to the 'code review' status. You were quick!

pacmessica commented on pipeline/pipeline_test.go:
```diff
@@ -123,6 +122,112 @@ func TestPipeline_SetStages(t *testing.T) {
+		runValidator.After(runParser)
+
+		runSequencer.After(runSyncer)
+		runAggregator.After(runSyncer)
```
runSequencer and  runAggregator should be after runValidator. woops.

silentlight commented:
As I am thinking more about it I think that we can simplify it even more.
When we create a new `default` pipeline stages are setup for us and the only thing that is left is adding tasks to the existing stage.
So we should really change the name of `SetStage` to something like `AddTasksToStage`
```
p.AddTasksToStage(pipeline.StageValidator, pipeline.SyncRunner(NewValidatorTask()))
```
Then when we want to create a custom pipeline, we could simply do:
```
p.AddStage(Stage {
  Name: pipeline.StageFetcher,
  TasksRunner: pipeline.SyncRunner(NewTask()),
})
```
and then for stages that we want to run concurrently we could have (variadic arg):
```
p.AddConcurrentStages(Stage {
  Name: "stage 1",
  TasksRunner: pipeline.SyncRunner(NewTask()),
}, Stage {
  Name: "stage 2",
  TasksRunner: pipeline.SyncRunner(NewTask()),
})
```
For creating a default pipeline we could to use combination of `AddStage` and `AddConcurrentStages` without specifying TasksRunner in the Stage struct. We will leave it to the user to use `AddTasksToStage` to add those tasks.
Let me know what you think about it?

pacmessica commented:
So then `p.stages` would be in the format `stages [][]StageName`, right? And then stages are run in the order in which they're added? I was considering that approach at first, but then I liked having the stages in a map so I came up with  the dependency approach. I think the stages in an array is more readable though. 

I like pipeline.New (for custom pipelines) and pipeline.Default method names better. Reason I didn't do it that way was because I didn't want to introduce breaking changes. But if that's not a concern, then I'll change them!

`AddTasksToStage` name is definitely more descriptive.

Do we want to prevent users from calling `AddStage`/`AddConcurrentStages` if it's a default pipeline? 

Thanks for the thoughtful feedback!

pacmessica commented:
> I saw it but I don't know the context or golang so I haven't got much to say about it :P
> 👍 for adding tests

that's about how I feel when going over the rails PRs 😅

silentlight commented:
@pacmessica, yeah I think the format you proposed looks right :)

I think we should prevent users from calling AddStage/AddConcurrentStages if it's a default pipeline. Maybe we should have a CustomPipeline interface?

pacmessica commented:
Implemented your suggestions 😄 

Changed method name `SetStage` to `SetStageRunner`. I didn't use `AddTasksToStage` because that name sounded like it was appending tasks to me.

I removed the `dependencies` dag from the pipeline struct, and changed stages format to `[][]*stage`. Default pipeline is a pipeline with all stages set in the default order with an empty stagerunner. Added a`Default` interface with `NewDefault` method and `Pipeline` interface with `New` method. I'm not sure about that naming though, do you prefer `CustomPipeline`/`NewCustomPipeline` for custom and `Pipeline`/`New` for default?

I still need to update the readme

silentlight commented on pipeline/pipeline.go:
```diff
@@ -55,56 +56,109 @@ type Options struct {
+}
+
+// Default implements a modular, multi-stage pipeline with set stages
+type Default interface {
```
Yeah, I think we can have 3 interfaces here:
```
type Pipeline interface {
  AddStageBefore(existingStageName StageName, name StageName, stageRunner StageRunner)
  AddStageAfter(existingStageName StageName, name StageName, stageRunner StageRunner)
  Start(ctx context.Context, source Source, sink Sink, options *Options) error
  Run(ctx context.Context, height int64, options *Options) (Payload, error)
}

type DefaultPipeline {
  Pipeline
  
  SetStageRunner(stageName StageName, stageRunner StageRunner)
}

type CustomPipeline {
  Pipeline

  AddStage(stageName StageName, stageRunner StageRunner)
  AddConcurrentStages(stageWithRunner ...StageWithRunner)
}
```

And then as you wrote we could have constructors like pipeline.NewDefault and pipeline.NewCustom.
How does it sound?

pacmessica replied:
sounds good to me!

silentlight commented on pipeline/pipeline.go:
```diff
@@ -55,56 +56,109 @@ type Options struct {
+// NewDefault creates a new default pipeline
+func NewDefault(payloadFactor PayloadFactory) Default {
+	p := New(payloadFactor)
+
```
I think something like type casting here will be enough since I am not sure we will ever use that emptyRunner anywhere
```
emptyRunner := StageRunnerFunc(func(context.Context, Payload, TaskValidator) error {
  return nil
})
```

pacmessica replied:
ahh yeah, good idea. thanks for pointing it out

silentlight commented on pipeline/stage.go:
```diff
@@ -43,6 +43,17 @@ func (s *stage) canRunTask(taskName string, options *Options) bool {
 	return true
 }
 
+func emptyRunner() StageRunner {
```
Left comment above about emptyRunner

pacmessica commented on pipeline/pipeline.go:
```diff
@@ -55,56 +56,109 @@ type Options struct {
 
-// Pipeline implements a modular, multi-stage pipeline
-type Pipeline struct {
+type StageWithRunner struct {
```
`Stage` is also an interface. So I'll change `stage` struct to `StageWithRunner` struct, or do you have a name preference?

silentlight replied:
I am not sure if we need StageWithRunner at all since we have NewStage?

silentlight commented on pipeline/pipeline.go:
```diff
@@ -55,56 +56,109 @@ type Options struct {
+	p.AddStage(StageFetcher, emptyRunner())
+	p.AddStage(StageParser, emptyRunner())
+	p.AddStage(StageValidator, emptyRunner())
+	p.AddConcurrentStages(StageWithRunner{StageSequencer, emptyRunner()}, StageWithRunner{StageAggregator, emptyRunner()})
```
Here instead of `StageWithRunner` can we use `NewStage`? This should fix the problem with naming convention above?

silentlight commented on pipeline/pipeline.go:
```diff
@@ -166,47 +229,29 @@ func (p *Pipeline) setupCtx(ctx context.Context) (context.Context, context.Cance
 
-	if err := p.runStage(StageCleanup, ctx, payload); err != nil {
-		return err
+		if err := p.runStage(ctx, stages[0], payload); err != nil {
```
I think that here we can simplify it a little bit doing:
```
if len(stages) == 1 {
  // runStage
} else if len(stages) > 1 {
  // runStagesConcurrently
} else {
   // log info that stage runner has not tasks?
}
```

What do you think?

pacmessica replied:
if `len(stages) == 0`, that means an empty array ([]*stage) was added to p.stages.  This is possible if they call 	`p.AddConcurrentStages()` with no arguments. More logging never seems like a bad idea to me, so I'll add it in.

If we want to log that a stagerunner has no tasks, then I think we should add it to the emptyrunner func, something like this:

```
func NewDefault(payloadFactor PayloadFactory) DefaultPipeline {
	p := new(payloadFactor)

	emptyRunner := func(name StageName) StageRunner {
		return StageRunnerFunc(func(context.Context, Payload, TaskValidator) error {
			logInfo(fmt.Sprintf("stage name %s not set up", name))
			return nil
		})
	}
	
	p.AddStage(StageSetup, emptyRunner(StageSetup))
       ...etc
```

Sounds good?

pacmessica replied:
Can also add a check to prevent user from adding empty stages, like so
```
func (p *pipeline) AddConcurrentStages(stages ...*stage) {
	if len(stages) == 0 {
		return
	}
	p.stages = append(p.stages, stages)
}
```

silentlight replied:
@pacmessica I just realized that we already have a logger for that in master branch
```
// canRunStage determines if stage can be ran
func (p *Pipeline) canRunStage(stageName StageName) bool {
	_, ok := p.stages[stageName]
	if !ok {
		logInfo(fmt.Sprintf("stage name %s not set up", stageName))
		return false
	}

	if p.options != nil && len(p.options.StagesBlacklist) > 0 {
		for _, s := range p.options.StagesBlacklist {
			if s == stageName {
				return false
			}
		}
		return true
	}

	return true
}
```

pacmessica replied:
yeah, the equivalent log now would be in the emptyrunner. I pushed the change.

pacmessica replied:
next steps? @silentlight 

a bunch of code will also change with this PR:
https://github.com/figment-networks/indexing-engine/pull/4/files

pacmessica commented:
waiting on a second approval. @sosedoff @pienkowb ?

sosedoff commented on pipeline/pipeline.go:
```diff
@@ -137,7 +207,7 @@ func (p *Pipeline) Start(ctx context.Context, source Source, sink Sink, options
 
 // Run run one-off pipeline iteration for given height
-func (p *Pipeline) Run(ctx context.Context, height int64, options *Options) (Payload, error) {
+func (p *pipeline) Run(ctx context.Context, height int64, options *Options) (Payload, error) {
```
Should this be a `uint64` ?

sosedoff replied:
I think i also had this type inconsistency in a few places now, so not a big deal, we can fix this later.

sosedoff commented on pipeline/pipeline.go:
```diff
@@ -55,56 +57,124 @@ type Options struct {
-func (p *Pipeline) SetStage(stageName StageName, stageRunner StageRunner) {
-	p.stages[stageName] = NewStage(stageName, stageRunner)
+// SetStageRunner sets up stagerunner for given stage
+func (p *pipeline) SetStageRunner(stageName StageName, stageRunner StageRunner) {
```
What happens when stage is not found?

pacmessica replied:
`SetStageRunner` is a method only available to the default pipeline, and in the default pipeline every single stage is set so it should always find the stage.

If a user tries to call this with a completely new stage that's not part of the default pipeline, eg.

`	p.SetStageRunner(
		pipeline.StageName("foo"),
		pipeline.AsyncRunner(...),
	)
`
Then nothing happens. No new stage should be added or set to a default pipeline, the user can only call `SetStageRunner` on existing stages in the pipeline.

I'll add a warning log though to cover that case

sosedoff commented on pipeline/pipeline.go:
```diff
@@ -166,47 +236,26 @@ func (p *Pipeline) setupCtx(ctx context.Context) (context.Context, context.Cance
 
 // runStagesConcurrently runs indexing stages concurrently
-func (p *Pipeline) runStagesConcurrently(ctx context.Context, payload Payload, stages []StageName) error {
+func (p *pipeline) runStagesConcurrently(ctx context.Context, payload Payload, stages []*stage) error {
```
I guess the stages argument could be defined as `stages ...*stage` ?

pacmessica replied:
This method is only being called in one place like this:
`if err := p.runStagesConcurrently(ctx, payload, stages); err != nil {`
where stages is of the type `[]*stage`. I think it looks cleaner the way it is now

I think it's

sosedoff replied:
🤙

pacmessica commented:
> LGTM overall, left a few notes around type references.

awesome, thank you! pushed one change
